src/dis6/VariableDatum.cpp

```diff
@@ -48,110 +45,65 @@ const char* VariableDatum::getVariableDatums() const
     return _variableDatums.data();
 }
 
-void VariableDatum::setVariableDatums(const char* x, const int length)
+void VariableDatum::setVariableDatums(const char* x, const unsigned int length)
 {
-    if(length < 0) {
-        // *** One should really be using unsigned for size values! f.e. std::size_t
-        std::cout << " The VariableDatum pointer size parameter is negative. Punting." << std::endl;
-    }
-
-    int byteLength = length; // why this copy?
-    _variableDatumLength = length * 8; // in bits
-
-	// Figure out padding
-    int chunks = byteLength / 8; // should be const, unsigned
-	int remainder = byteLength % 8;
-	if(remainder != 0)
-		chunks++;
-    _arrayLength = chunks * 8;
-
-    int padding =  8 - remainder;
-
-    // .resize() might (theoretically) throw. want to catch? : what to do? zombie datum?
-    // negative "length" would be a disaster : signed->unsigned->huge allocation
-    if(_variableDatums.size() < length)
-        _variableDatums.resize(length);
-
-    for(int i = 0; i < length; i++)
+    if(_variableDatums.size() != length)
     {
-        _variableDatums[i] = x[i];
-    }
-    for(int i = length; i < _variableDatums.size(); i++)
-    {
-        _variableDatums[i] = 0;
+        // Standard library will set new size to length but may not unallocate memory
+        // As a result, resizing to a smaller size doesn't mean we'll waste time allocating
+        // when a larger datum arrives.
+        _variableDatums.resize(length); 
     }
+    std::memcpy(&_variableDatums[0], x, length);
 }
 
 void VariableDatum::marshal(DataStream& dataStream) const
 {
     dataStream << _variableDatumID;
-    dataStream << _variableDatumLength;
+    dataStream << _variableDatums.size();
 
-    for(size_t idx = 0; idx < _arrayLength; idx++)
+    for(size_t idx = 0; idx < _variableDatums.size(); idx++)
     {
         dataStream << _variableDatums[idx];
     }
-
 }
 
 void VariableDatum::unmarshal(DataStream& dataStream)
 {
     dataStream >> _variableDatumID;
-    dataStream >> _variableDatumLength;
+
+    size_t incoming_length;
+
+    dataStream >> incoming_length;
 	
-    int byteLength = _variableDatumLength / 8;
+    int byteLength = incoming_length / 8;
 	int chunks = byteLength / 8;
 	if(byteLength % 8 > 0)
```

Comment:
I think line 104 actually should have been `if(_variableDatumLength % 8 > 0)`, because it is meant to check if the Variable Datum fits into whole Bytes - or rather if it checks that it doesn't, and then adds the extra byte to cover the remaining bits. 
Therefore in the new file, line 81 should be `if(incoming_length % 8 > 0)`.

Reply:
Scratch that. I didn't realise that the chunks ar 8 bytes (64 bits)